### PR TITLE
RunThreaded changes.

### PR DESCRIPTION
**Pull request by skschneider**
Returns function return values.
Schedules calls in the order of the args lists (previously reverse).
Preserves interruptability.

### CODE REVIEW

**Comment by cmccoy** [comment]
Looks nice to have. Any chance of using `ThreadPoolExecutor` from `concurrent.futures`? If possible, it would be nice to have a single function to execute tasks in parallel (that takes an `Executor`).

**Comment by skschneider** [comment]
I encountered two problems with ThreadPoolExecutor: doesn't support daemon threads / custom thread types, and isn't interruptable in Python 2.7. If you know a solution for these, I agree it would be preferable to use that class.

**Comment by cmccoy** [comment]
I believe the threads spawned [are daemons](https://hg.python.org/cpython/file/3.5/Lib/concurrent/futures/thread.py#l128), but I don't know of any way to handle custom thread types / interruption in 2.7. Thanks for the explanation.

File: perfkitbenchmarker/vm_util.py [hunk 1]
```diff
@@ -172,32 +173,29 @@ def GetSshOptions(ssh_key_filename):
...
-    """Modified join to raise an exception if needed."""
-    super(ThreadWithExceptions, self).join(*args, **kwargs)
-    if self.exception:
-      raise errors.VmUtil.ThreadException(self.exception)
+  Args:
+    target_arg_tuple: (target, args, kwargs) tuple containing the function to
+        call and the arguments to pass it.
+    call_id: int. Index corresponding to the call in the thread_params argument
+        of RunThreaded.
+    queue: Queue. Receives a (call_id, result, stacktrace) tuple. If the called
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Maybe introduce a namedtuple for these?

File: perfkitbenchmarker/vm_util.py [hunk 2]
```diff
@@ -238,42 +240,55 @@ def RunThreaded(target, thread_params, max_concurrent_threads=200):
...
-      args, kwargs = thread_params.pop()
-      thread = ThreadWithExceptions(target=target, args=args, kwargs=kwargs)
-      threads.append(thread)
+      # Using a timeout makes this wait interruptable.
+      call_id, result, stacktrace = queue.get(block=True, timeout=1000)
+    except Queue.Empty:
+      continue
+    results[call_id] = result
+    if stacktrace:
+      msg = 'Exception occured in call {0}:{1}{2}'.format(
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Could you log the args?

File: perfkitbenchmarker/vm_util.py [hunk 3]
```diff
@@ -238,42 +240,55 @@ def RunThreaded(target, thread_params, max_concurrent_threads=200):
     return
 
   if not isinstance(thread_params[0], tuple):
-    thread_params = [((arg,), {}) for arg in thread_params]
+    target_arg_tuples = [(target, (arg,), {}) for arg in thread_params]
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Consider moving the logic below into a function that accepts `target_arg_tuples`, so multiple functions can be run concurrently as in the multiprocess case.

File: perfkitbenchmarker/vm_util.py [hunk 4]
```diff
@@ -238,42 +240,55 @@ def RunThreaded(target, thread_params, max_concurrent_threads=200):
...
+                         for args, kwargs in thread_params]
+
+  queue = Queue.Queue()
+  log_context = log_util.GetThreadLogContext()
+  max_concurrent_threads = min(max_concurrent_threads, len(target_arg_tuples))
+  results = [None] * len(target_arg_tuples)
+  error_strings = []
+  for call_id in xrange(max_concurrent_threads):
+    target_arg_tuple = target_arg_tuples[call_id]
+    thread = threading.Thread(
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: consider naming the thread.

File: perfkitbenchmarker/vm_util.py (hunk 3 above)
Comment (skschneider) [inline perfkitbenchmarker/vm_util.py]: Would have been my next PR. Will add it to this one.

File: perfkitbenchmarker/vm_util.py (hunk 4 above)
Comment (skschneider) [inline perfkitbenchmarker/vm_util.py]: I think we've used default naming so far. RunThreaded can be called from within a function invoked by RunThreaded, so a name based on the call_id in this function wouldn't be good. An ideal name might be based on what the thread's role is, but the implementation for that would be complicated. I think it's best to delay on that and add it as a separate feature.

Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: SGTM.

File: perfkitbenchmarker/vm_util.py (hunk 1 above)
Comment (skschneider) [inline perfkitbenchmarker/vm_util.py]: Done.

File: perfkitbenchmarker/vm_util.py (hunk 2 above)
Comment (skschneider) [inline perfkitbenchmarker/vm_util.py]: Done.

File: perfkitbenchmarker/vm_util.py (hunk 3 above)
Comment (skschneider) [inline perfkitbenchmarker/vm_util.py]: Done.

**Comment by cmccoy** [comment]
test this please.

**Comment by cmccoy** [comment]
test this please.

**Comment by cmccoy** [comment]
LGTM.

**Pull request merged**